sw/device/lib/dif/README.md

```diff
@@ -50,22 +50,44 @@ memory-mapped hardware, and modifying processor CSRs.
   DIFs must not call DIFs in other DIF libraries.
 * DIF library headers must be polyglot headers for C and C++.
 
-### DIF Guidance
+### DIF API Guidance
 
-The following rules must be followed by public DIF functions (those declared in
-the DIF library's header file). Internal DIF functions (those declared `static`
-and not declared in the DIF library's header file) should follow these rules but
-there are some relaxations of these rules for them described at the end.
+The following rules specify the basic API that each DIF must conform to. These
+rules specify the names of types, constants, and functions that each DIF must
+define for providing certain kinds of non-device-specific functionality (such as
+initializing handles or managing interrupts).
 
-* DIF declarations must match their definitions exactly.
-  * Scalar arguments must not be declared `const` or `volatile` (cv-qualified)
-    in DIF signatures.
+Notational caveats:
+* The token `<ip>` is the "short IP name" of the peripheral,
+  in `snake_case` or `PascalCase` as is appropriate.
+* The parameter name `handle` is not normative, and DIF libraries are free to
+  choose a different, but consistent, name for it.
+* All functions below are assumed to return `dif_<ip>_result_t` or
+  `dif_<ip>_<operation>_result_t`, so their return values are specified as
+  `result_t`.
+* Unless otherwise noted, all symbols mentioned below are required.
+
+[`dif_template.h.tpl`](./dif_tempalte.h.tpl) provides a starting point for
+building a DIF library that conforms to this guidance.
+
+#### Base Types
 
-* DIFs must use enum return codes rather than booleans for reporting errors. If
-  a DIF can either error or instead produce a value, it must return an enum
-  return code, and use an out-parameter for returning the produced value.
-  * DIF libraries must have a general return code enum called
-    `dif_<ip>_result_t`. It must define the following constants (in this order):
+The following basic types are expected to be provided by all DIFs (unless
+otherwise specified).
+
+* `dif_<ip>_t` is a type representing a handle to the peripheral. Its fields
+  are private implementation details, and should not be read or written to by
+  clients. This type is always passed by `const` pointer, except when it is
+  being initialized (see `dif_<ip>_init()` below).
```

Comment:
We have DIFs that keep some software state in their `dif_<ip>_t` handles (SPI Dev, USB Dev) - downgrade this from `always` to `usually`.